src/tactic/scc.lean

```diff
@@ -12,66 +12,174 @@ of implications.
+and edges are proofs that the source implies the target. The strongly connected components are therefore
+sets of propositions that are each equivalent to each other.
+
+The resulting strongly connected components are encoded in a disjoint set data structure to facilitate
+the construction of proof of equivalence between two arbitrary members of an equivalence class.
+
+## Possible generalizations
+
+Instead of reasoning about implications and equivalence, we could generalize the machinery to
+reason about arbitrary partial orders.
+
+## References
+
+ * Tarjan, R. E. (1972), "Depth-first search and linear graph algorithms",
+   SIAM Journal on Computing, 1 (2): 146–160, doi:10.1137/0201010
+ * Dijkstra, Edsger (1976), A Discipline of Programming, NJ: Prentice Hall, Ch. 25.
+ * https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+## Tags
+
+graphs, tactic, strongly connected components, disjoint sets
+-/
 
 namespace tactic
 
-meta def closure := ref (expr_map (expr × expr))
+/--
+`closure` implements a disjoint set data structure using path compression
+optimization. For the sake of the scc algorithm, it also stores the preorder
+numbering of the equivalence graph of the local assumptions.
+
+The `expr_map` encodes a directed forest by storing for every non-root
+node, a reference to its parent and a proof of equivalence between
+that node's expression and its parent's expression. Given that data
+structure, checking that two nodes belong to the same tree is easy and
+fast by following the parent reference repeated up to a root. If they
+both have the same root, they belong to the same tree, i.e. their
+expressions are equivalent. The proof of equivance can be formed by
+composing the proofs along the edges of the paths to the root.
+
+More concretely, if we ignore preorder numbering, the set
+`{ {e₀,e₁,e₂,e₃}, {e₄,e₅} }` is represented as:
+
+```
+e₀ → ⊥      -- no parent, i.e. e₀ is a root
+e₁ → e₀, p₁ -- with p₁ : e₁ ↔ e₀
+e₂ → e₁, p₂ -- with p₂ : e₂ ↔ e₁
+e₃ → e₀, p₃ -- with p₃ : e₃ ↔ e₀
+e₄ → ⊥      -- no parent, i.e. e₄ is a root
+e₅ → e₄, p₅ -- with p₅ : e₅ ↔ e₄
+```
+
+We can check that `e₂` and `e₃` are equivalent by seeking the root of
+the tree of each. The parent of `e₂` is `e₁`, the parent of `e₁` is
+`e₀` and `e₀` does not have a parent, and thus, this is the root of its root.
+The parent of `e₃` is `e₀` and it's also the root, the same as for `e₃` and
+they are therefore equivalent. We can build a proof of that equivalence by using
+transitivity on `p₂`, `p₁` and `p₃.symm` in that order.
+
+Similarly, we can discover that `e₂` and `e₅` aren't equivalent.
+
+A description of the path compression optimization can be found at:
+https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Path_compression
+
+-/
+meta def closure := ref (expr_map (ℕ ⊕ (expr × expr)))
 
 namespace closure
 
 meta def mk_closure {α} : (closure → tactic α) → tactic α :=
```

Comment:
doc string

Reply:
It looks like Scott and I reviewed at exactly the same time yesterday. I'm going to resolve half of these comments!